PyDynamic/model_estimation/fit_filter.py

```diff
@@ -87,92 +91,224 @@ def _fitIIR(
-        g2 = grpdelay(b, astab, Fs)[0]
-        tau = np.ceil(tau + np.median(g2 - g1))
+    # Initialize the warning message in case the final filter will still be unstable.
+    warning_unstable = "CAUTION - The algorithm did NOT result in a stable IIR filter!"
 
-        b, a = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=False)
-        if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-            astab = mapinside(a)
+    # Prepare frequencies, fitting and stabilization parameters.
+    w = 2 * np.pi * f / Fs
+    Ns = np.arange(0, max(Nb, Na) + 1)[:, np.newaxis]
+    E = np.exp(-1j * np.dot(w[:, np.newaxis], Ns.T))
+    as_and_bs = np.empty((mc_runs, Nb + Na + 1))
+    taus = np.zeros((mc_runs,), dtype=int)
+    stab_iters = np.zeros((mc_runs,), dtype=int)
+
+    # Conduct the Monte Carlo runs or in case we did not have uncertainties execute
+    # just once the actual algorithm.
+    for mc_run in range(mc_runs):
+        # Conduct actual fit.
+        b_i, a_i = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=inv)
+
+        # Determine if the computed filter already is stable.
+        unstable = not check_stability(a_i)
+
+        # In case the user specified not to check for stability and stabilize the filter
+        # if necessary, we just return the result so far and inform about hte result.
+        if justFit:
+            if verbose:
+                sos = np.sum(np.abs((dsp.freqz(b_i, a_i, w)[1] - Hvals) ** 2))
+                print(
+                    f"LSIIR: Fitting "
+                    f"{'' if UHvals is None else f'for MC run {mc_run} '}"
+                    f"finished. No stabilization requested"
+                    f". {warning_unstable if unstable else ''} "
+                    f"Final sum of squares = {sos}"
+                )
+            # Since no stabilization was done, we return tau = 0 regardless of the
+            # desired initial estimate of the time delay provided for the stabilization.
+            taus[mc_run] = 0
+            continue
         else:
-            stable = True
-        run = run + 1
+            # Otherwise we initialize the time delay for the current Monte Carlo run.
+            taus[mc_run] = tau
+
+        # Initialize counter which we use to report about required iteration count.
+        current_stab_iter = 0
+        # Stabilize filter coefficients with a maximum of 50 iterations.
+        while unstable and current_stab_iter < max_stab_iter:
+            # Compute appropriate time delay for the stabilization of the filter.
+            a_stab = mapinside(a_i)
+            g_1 = grpdelay(b_i, a_i, Fs)[0]
+            g_2 = grpdelay(b_i, a_stab, Fs)[0]
+            taus[mc_run] = np.ceil(taus[mc_run] + np.median(g_2 - g_1))
+
+            # Conduct stabilization step through time delay.
+            b_i, a_i = _fitIIR(Hvals, taus[mc_run], w, E, Na, Nb, inv=inv)
+
+            # Prepare abortion in case filter is stable.
+            if check_stability(a_i):
+                unstable = False
```

Comment:
I think we should quickly check if this is really faster than `unstable = not check_stability(a_i)`.

Reply:
Turns out the conditional itself without any action takes about as long as  an assigment:

```python
>>> if_clause = """
... if True:
...     pass
... """
>>> timeit(if_clause)
0.11669498800074507
>>> timeit("unstable = False")
0.13944279300085327
```

Thus, it appears reasonable to just assign the result of `not check_stability()` to the variable!

Reply:
This has been done in 8d1f962ff994a258fac6a8694cfa798abb7f2066.